### PR TITLE
[receiver/snowflake] add metrics

### PR DESCRIPTION
**Pull request by shalper2**
**Description:** 
Defined metrics in metadata.yaml and ran mdatagen to create the metadata package for use in receiver component. Metrics are designed for parity with existing [smartagent](https://github.com/splunk/observability-content-contrib/blob/main/dashboards-and-dashboard-groups/snowflakedb/Configuration/snowflake-metrics.yaml). Dimensions have been translated into attributes with the same column values associated. The smartagent definition does not include data types so care was taken to apply applicable types to each metric and attribute field.

**Link to tracking Issue:** [14754](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/14754)

**Testing:** 
simple functional testing

**Documentation:** 
generated using [mdatagen](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/cmd/mdatagen)

### CODE REVIEW

File: receiver/snowflakereceiver/documentation.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+| error.message | Error message if present |  |
+| execution.status | execution status |  |
+| is.success | Login status |  |
+| pipe.name | name of snowpipe |  |
+| query.type | type of query performed |  |
+| reported.client.type | Client type used for attempt |  |
+| schema.name | name of schema |  |
+| service.type | Service type |  |
+| user.name | username in query |  |
+| warehouse.name | name of Warehouse in query |  |
```
Comment (dmitryax) [inline receiver/snowflakereceiver/documentation.md:83]: Why do we have warehouse name both in metric attributes and resource attributes?

File: receiver/snowflakereceiver/documentation.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+| snowflake.warehouse.name | The name of the warehouse being used | String |
+
+## Metric attributes
+
+| Name | Description | Values |
+| ---- | ----------- | ------ |
+| database.name | name of database (default is snowflake) |  |
+| error.message | Error message if present |  |
+| execution.status | execution status |  |
+| is.success | Login status |  |
```
Comment (dmitryax) [inline receiver/snowflakereceiver/documentation.md:76]: Dots are usually used in the attribute/metrics names to separate namespace from the entity. I don't `is` is a namespace. This should `is_success` or maybe just `success`?

File: receiver/snowflakereceiver/documentation.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+| pipe.name | name of snowpipe |  |
+| query.type | type of query performed |  |
+| reported.client.type | Client type used for attempt |  |
+| schema.name | name of schema |  |
+| service.type | Service type |  |
+| user.name | username in query |  |
+| warehouse.name | name of Warehouse in query |  |
+| warehouse.size | size of Warehouse in query |  |
```
Comment (dmitryax) [inline receiver/snowflakereceiver/documentation.md:84]: this seems like a non-identifying resource attribute, no?

Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md:84]: you're absolutely right these should all be underscores and not dots. I did a find and replace when I was defining these metrics and caught a little too much.

File: receiver/snowflakereceiver/documentation.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+# snowflakereceiver
+
+## Metrics
+
+These are the metrics available for this scraper.
+
+| Name | Description | Unit | Type | Attributes |
+| ---- | ----------- | ---- | ---- | ---------- |
+| **snowflake.billing.cloud_service.sum** | cloud services sum | credits | Gauge(Double) | <ul> <li>service_type</li> </ul> |
```
Comment (dmitryax) [inline receiver/snowflakereceiver/documentation.md:11]: Few comments for all the metrics:

1. Please make sure that all sum metrics are monotonic or non-monotonic sums, not gauges. So it's clear that they can be aggregated over the all the attributes by sum.
2. Please improve the descriptions: provide some idea about what the metric represents, not just repeating its name. And description should be a sentence starting with a capital letter.
3. Please don't make all the metrics enabled by default. Is seems to be a lot of data. As a rule of thumb, metrics enabled by default should provide critical information about heathiness and performance of a monitored system. Other metrics should be optional. All `snowflake.billing` metrics are good candidate for optional metrics. And maybe it's even better to start with a smaller set of essential metrics and add more later on?
4. Enclose all countable units into {brackets}, e.g. credits -> {credits}.

File: receiver/snowflakereceiver/documentation.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+| **snowflake.billing.total_credit.sum** | billing total sum | credits | Gauge(Double) | <ul> <li>service_type</li> </ul> |
+| **snowflake.billing.virtual_warehouse.sum** | compute credits used sum |  | Gauge(Double) | <ul> <li>service_type</li> </ul> |
+| **snowflake.billing.warehouse.cloud_service.sum** | ware billing for cloud service | credits | Gauge(Double) | <ul> <li>warehouse_name</li> </ul> |
```
Comment (dmitryax) [inline receiver/snowflakereceiver/documentation.md:14]: Is this metrics based on the same data as snowflake.billing.cloud_service.sum just broken down by different attribute? Can we use just `snowflake.billing.cloud_service.sum` with both `warehouse_name` and `service_type` attributes?

**Review by dmitryax — COMMENTED** [review]
Sorry for the delay. I was pretty swamped recently. Will be responding quickly going forward

File: receiver/snowflakereceiver/documentation.md (hunk 4 above)
Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md:11]: Was on PTO going through this now! So sorry for the delay...

1. There is a naming conflict here. In the original snowflake monitoring tool the word sum is meant to represent the aggregate total as reported by the sql query which actually contains the results. This table entry is not aggregated by the observability tool, but rather displayed as it is reported. Should the name of these sum metrics be changed to reflect that or would keeping with previous convention be better?
2. Heard.
3. Heard. I was not initially aware that this was the norm, will fix this moving forwards.
4. Heard.

File: receiver/snowflakereceiver/documentation.md (hunk 5 above)
Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md:14]: No so all of the data represented in each metric is actually the result of a query run in snowflake (this is how they do reporting) so they're all actually distinct values.

File: receiver/snowflakereceiver/documentation.md (hunk 4 above)
Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md]: I went ahead and changed all metrics named "sum" to "total" for clarity

File: receiver/snowflakereceiver/documentation.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    enabled: <true|false>
+```
+
+## Resource attributes
+
+| Name | Description | Type |
+| ---- | ----------- | ---- |
+| snowflake.account.name | The name of the snowflake account being monitored. | String |
+| snowflake.username | The name of the snowflake user account being used. | String |
+| snowflake.warehouse.name | The name of the warehouse being used. | String |
```
Comment (dmitryax) [inline receiver/snowflakereceiver/documentation.md:67]: I see these resource attributes are added to some metrics as metrics attributes. We should have either resource attributes or metric attributes, not both.

Resource attributes are supposed to be applied to all of the metrics. If we cannot apply them to all metrics, we should keep metric attributes instead and remove these resource attributes.

**Review by dmitryax — COMMENTED** [review]
Please rebase and run `make generate`. I recently changed the generated documentation

File: receiver/snowflakereceiver/documentation.md (hunk 6 above)
Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md:67]: This requires a little context: The way that snowflake reports these statistics is by running a query. In order to do this via a client call however, an organization is required to define a warehouse (compute resources) a database to search (the default location of the logs is the SNOWFLAKE database) and a user with the appropriate permissions. 

The resource attributes as defined here are meant to reflect those values (as they would be defined in the config) and are not necessarily the same as the metric attributes with similar names. They are also the same for for every metric being generated by any particular collector.

I'm going to update the descriptions to better reflect this sentiment, but if you still don't think that these attributes should be reported let me know!

Comment (dmitryax) [reply receiver/snowflakereceiver/documentation.md:67]: Ok, in order to decide if we need to include these resource attributes we need to answer couple of questions:

- Can the user have another warehouse with different permissions to query the same data? 
- How important that information is to be included in the generated metrics? Does the presence of these resource attributes justify possible confusion with metric attributes? Keep in mind that many exporters don't pass the description field and resource attributes are merged with metrics attributes, e.g. that's the case for signalfx exporter.

Comment (dmitryax) [reply receiver/snowflakereceiver/documentation.md:67]: And resource attributes are supposed to be applied to all the metrics as static attributes, but AFAIU these ones are applied to some "proxy" warehouse that just stores the usage statistics about all other warehouses. Please correct me if my understanding is wrong. If that's the case, I don't think `snowflake.username` and `snowflake.warehouse.name` should be added to resource attributes. But `snowflake.account.name` is a good example of resource attribute because it can be clearly applied to all the metrics.

File: receiver/snowflakereceiver/documentation.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+# snowflakereceiver
+
+## Metrics
+
+These are the metrics available for this scraper.
+
+| Name | Description | Unit | Type | Attributes |
+| ---- | ----------- | ---- | ---- | ---------- |
+| snowflake.billing.cloud_service.total | Reported total credits used in the cloud service. | {credits} | Gauge(Double) | <ul> <li>service_type</li> </ul> |
```
Comment (dmitryax) [inline receiver/snowflakereceiver/documentation.md:11]: the documentation is still in the old format. Please make sure to run `make generate` after rebasing

File: .chloggen/ottl-merge-maps.yaml~merged [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+<<<<<<< HEAD
+<<<<<<< HEAD:.chloggen/ottl-merge-maps.yaml
+=======
+>>>>>>> 3087cc9a0e8e732303bf78d14c668e063d42e6df
```
Comment (dmitryax) [inline .chloggen/ottl-merge-maps.yaml~merged:8]: Please resolve the conflicts

Comment (dmitryax) [reply .chloggen/ottl-merge-maps.yaml~merged:8]: Oh this is actually some merge artifact and shouldn't be included in the PR

File: receiver/snowflakereceiver/documentation.md (hunk 6 above)
Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md:67]: So just to clarify [warehouses](https://docs.snowflake.com/en/user-guide/warehouses-overview.html) are what snowflake calls their compute units. All queries must be associated with a warehouse, and they really don't have much to do with storing data (i know...). So when snowflake.warehouse.name is applied what its really referring to is the compute unit associated with the stats queries we're scraping. 

However! The more I think about it the more I think it might not really matter what the compute resource is named, so I'm all in favor of dropping `snowflake.warehouse.name` altogether. That said, I can imagine scenarios where wanting to know the username and account names associated with these metrics. Although if you're really passionate about removing username I won't argue.

File: .chloggen/ottl-merge-maps.yaml~merged (hunk 8 above)
Comment (dmitryax) [reply .chloggen/ottl-merge-maps.yaml~merged:8]: it's still there :)

Comment (shalper2) [reply .chloggen/ottl-merge-maps.yaml~merged]: ok, ok, NOW its gone (I think)

File: .chloggen/unreleased/dd-exporter-duplicate-tags.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Doesn't append duplicate ddtags on each log submission leading to 414 API errors."
+
+# One or more tracking issues related to the change
+issues: [16380]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (dmitryax) [inline .chloggen/unreleased/dd-exporter-duplicate-tags.yaml:16]: Please make sure that no unrelated changes are included

File: .chloggen/ottl-merge-maps.yaml~merged (hunk 8 above)
Comment (dmitryax) [reply .chloggen/ottl-merge-maps.yaml~merged]: It's gone but there are other unrelated changelog item added after rebasing :)

File: receiver/snowflakereceiver/documentation.md (hunk 6 above)
Comment (dmitryax) [reply receiver/snowflakereceiver/documentation.md]: > The original justification for these resource attributes was to handle the case where multiple, separate snowflake deployments (say from two different business units within the same company) are being monitored in the same tool by the metric users.

In this case, each different business unit will be uniquely identified by an account name, right?

Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md]: Yea exactly! Having a user to chase down could also be useful but yes ultimately I think account name should be enough.

Comment (shalper2) [reply receiver/snowflakereceiver/documentation.md]: Ok great! Anything else?

Comment (dmitryax) [reply receiver/snowflakereceiver/documentation.md]: I see you updated the metadata.yaml, but you need to run `make generate` to update the generated files . Also please remove the unrelated changes as I mentioned in another comment

File: .chloggen/unreleased/dd-exporter-duplicate-tags.yaml (hunk 9 above)
Comment (shalper2) [reply .chloggen/unreleased/dd-exporter-duplicate-tags.yaml]: Not sure how that got in there! sorry

File: cmd/otelcontribcol/go.mod [hunk 10]
```diff
@@ -1,1074 +0,0 @@
...
-replace github.com/cloudflare/cloudflare-go => github.com/cyriltovena/cloudflare-go v0.27.1-0.20211118103540-ff77400bcb93
-
-// using fork for custom dialer support for loki - see https://github.com/bradfitz/gomemcache/pull/86
-replace github.com/bradfitz/gomemcache => github.com/grafana/gomemcache v0.0.0-20220812141859-1e3ae89e91a7
-
-// fork containing a line-buffered logger which should improve logging performance for loki
-replace github.com/go-kit/log => github.com/dannykopping/go-kit-log v0.2.2-0.20221002180827-5591c1641b6b
-
-// see https://github.com/mattn/go-ieproxy/issues/45
-replace github.com/mattn/go-ieproxy => github.com/mattn/go-ieproxy v0.0.1
```
Comment (dmitryax) [inline cmd/otelcontribcol/go.mod:1074]: Another merge artifact :) Please don't remove unrelated files

**Review by dmitryax — APPROVED** [review]
